### PR TITLE
feat: Query the catalog with PartitionHashId if you have it, otherwise with PartitionId

### PR DESCRIPTION
**Pull request by carols10cents**
Connects to https://github.com/influxdata/idpe/issues/17476.

Extracted from https://github.com/influxdata/influxdb_iox/pull/8068 if you're interested in seeing the direction in which this work is heading.

This PR changes these three catalog methods to accept a [`TransitionPartitionId`](https://github.com/influxdata/influxdb_iox/blob/7e30c91cebc32be665b360c5ab0feed09c038882/data_types/src/partition.rs#L10-L20) (which is a `PartitionHashId` if it's available and a `PartitionId` if not) instead of always a `PartitionId` as a means of identifying a partition. Each method change has its own commit in this PR:

- Looking up a partition by ID (done with a new function `partition_lookup` because there are methods `PartitionRepo::get_by_id` and `PartitionRepo::get_by_hash_id` already implemented per catalog impl and the `partition_lookup` logic doesn't depend on what catalog impl you're using; there are some places we still want to explicitly look up by `PartitionId` with `get_by_id`)
- `PartitionRepo::cas_sort_key`
- `ParquetFileRepo::list_by_partition_not_to_delete`

The remaining catalog methods that take `PartitionId`s as parameters are:

- `PartitionRepo::get_by_id`, mentioned above
- Methods used by the compactor:
  - `PartitionRepo::record_skipped_compaction`
  - `PartitionRepo::get_in_skipped_compaction`
  - `PartitionRepo::delete_skipped_compactions`

Because the compactor is querying the catalog anyway (in other words, we're not trying to eliminate those catalog queries like we are trying to eliminate the catalog queries in the ingester), I'm planning on leaving the compactor code to use `PartitionId` until the point at which we want to get rid of `PartitionId` completely.

### CODE REVIEW

File: iox_catalog/src/postgres.rs [hunk 1]
```diff
@@ -1458,22 +1472,38 @@ RETURNING id;
...
-            r#"
+        // This `match` will go away when all partitions have hash IDs in the database.
+        let query = match partition_id {
+            TransitionPartitionId::Deterministic(hash_id) => sqlx::query_as::<_, ParquetFile>(
+                r#"
+SELECT id, namespace_id, table_id, partition_id, partition_hash_id, object_store_id, min_time,
+       max_time, to_delete, file_size_bytes, row_count, compaction_level, created_at, column_set,
+       max_l0_created_at
+FROM parquet_file
+WHERE parquet_file.partition_hash_id = $1
```
Comment (crepererum) [inline iox_catalog/src/postgres.rs:1485]: We need a new partial index. See

https://github.com/influxdata/influxdb_iox/blob/main/iox_catalog/migrations/20230530132809_add_parquet_file_partition_and_deleted_index.sql

Please add this in an extra PR prior to merging this PR here so the perf impact isn't too harsh.

**Review by crepererum — CHANGES_REQUESTED** [review]
Needs PG index. Otherwise the PR looks pretty straight forward.

**Review by domodwyer — APPROVED** [review]
Ditto on the index, but looks good! 

Might be worth adding the index `CONCURRENTLY` now that it's do-able - we've had a few weeks of populating this column now.

**Comment by carols10cents** [comment]
Index added, ready for re-review! ❤️

**Comment by carols10cents** [comment]
Ok *now* this is ready for re-review.

**Review by domodwyer — APPROVED** [review]

**Pull request merged**